File: ruby/binary/binary.rb
module Binary
  BINARY_STRING = /\A[01]+\z/.freeze
  private_constant :BINARY_STRING
  BASE = 2
  private_constant :BASE
  DECIMAL_SEED = 0
  private_constant :DECIMAL_SEED

  module_function

  def to_decimal(string)
    raise ArgumentError unless string.match?(BINARY_STRING)

    string.chars.reverse.each.with_index.reduce(
      DECIMAL_SEED, &method(:add_decimal_value)
    )
  end

  def add_decimal_value(decimal, (bit, exponent))
    decimal + bit.to_i * BASE**exponent
  end
  private_class_method :add_decimal_value
end

module BookKeeping
  VERSION = 3
end

File: elixir/binary/lib/binary.ex
defmodule Binary do
  @binary_string ~r/\A[01]+\z/
  @base 2

  @doc """
  Convert a string containing a binary number to an integer.

  On errors returns 0.
  """
  @spec to_decimal(String.t()) :: non_neg_integer
  def to_decimal(string) do
    if String.match?(string, @binary_string) do
      string
      |> String.codepoints()
      |> Enum.reverse()
      |> Enum.with_index()
      |> Enum.reduce(0, &add_decimal_value/2)
    else
      0
    end
  end

  defp add_decimal_value({bit, index}, acc) do
    acc + String.to_integer(bit) * :math.pow(@base, index)
  end
end
